bip-unknown-var-budget-script.mediawiki

```diff
@@ -0,0 +1,722 @@
+1. The time taken to verify a Schnorr signature
+2. The time taken for simple operations on maximal (4000000 byte) stack entries
+
+From this, we can establish the value of a single varop budget, and determine how much budget would be equivalent to the already-established worst-case signature validation (80,000 Schnorr signature validations).  While CHECKSIG can be done in parallel, it also involves hashing, which is not taken into account here (the worst-case being signficant slower than the signature validations themselves).
+
+This would be the total budget for a maximal (4,000,000 weight) block.
+
+===SUCCESS Opcodes===
+
+The following opcodes are renamed OP_SUCCESSx, and cause validatation to immediately succeed:
+
+* OP_1NEGATE = OP_SUCCESS79
+* OP_NEGATE = OP_SUCCESS143
+* OP_ABS = OP_SUCCESS144
+
+====Rationale====
+
+Negative numbers are not natively supported in v2 Tapscript.  Arbitrary precision makes them difficult to manipulate and negative values are not used meaningfully in bitcoin transactions.
+
+===Enabled Opcodes===
+
+Fifteen opcodes which were removed in v0.3.1 are re-enabled in v2 Tapscript.
+
+If there are less than the required number of stack elements, these opcodes fail.  Equivalently, a requirement to pop off the stack which cannot be satisfied causes the opcode to fail.
+
+Splice opcodes:
+
+{|
+! Opcode
+! Value
+! Required Stack Elements
+! Definition
+|-
+|OP_CAT
+|126
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# Append B to A.
+# Push A onto the stack.
+|-
+|OP_SUBSTR
+|127
+|3
+|
+# Pop LEN off the stack.
+# Pop BEGIN off the stack.
+# Pop A off the stack.
+# Remove BEGIN bytes from the front of A (all bytes if BEGIN greater than length of A).
+# If length(A) is greater than value(LEN), truncate A to length value(LEN).
+# Push A onto the stack.
+|-
+|OP_LEFT
+|128
+|2
+|
+# Pop OFFSET off the stack.
+# Pop A off the stack.
+# If length(A) is greater than value(OFFSET), truncate A to length value(OFFSET).
+# Push A onto the stack.
+|-
+|OP_RIGHT
+|129
+|2
+|
+# Pop OFFSET off the stack.
+# Pop A off the stack.
+# Remove value(OFFSET) bytes from the front of A, or all bytes if value(OFFSET) is greater than the length(A).
+# Push A onto the stack.
+|}
+
+Bit operation opcodes:
+
+{|
+! Opcode
+! Value
+! Required Stack Elements
+! Definition
+|-
+|OP_INVERT
+|131
+|1
+|
+# Pop A off the stack.
+# For each byte in A, replace it with that byte bitwise XOR 0xFF (i.e. invert the bits)
+# Push A onto the stack.
+|-
+|OP_AND
+|132
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# If B is longer than A, swap B and A.
+# For each byte in A (the longer operand): bitwise AND it with the equivalent byte in B (or 0 if past end of B)
+# Push A onto the stack.
+|-
+|OP_OR
+|133
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# If B is longer than A, swap B and A.
+# For each byte in B (the shorter operand): bitwise OR it with the equivalent byte in A.
+# Push A onto the stack.
```

Comment:
Why does AND operate on the longer operand, and OR and XOR operatoe on the shorter operand?

Reply:
In all three cases, you modify the longer operand; but in the case of OR/XOR, you only need to iterate over the shorter operand, because any trailing bytes in the longer will remain unchanged. In the case of AND, you need to iterate over all the bytes in the longer operand, because any trailing bytes need to end up as zeroes.

Reply:
That makes sense. Maybe worth a footnote?